### ISSUE TITLE
Avoid returning `^a.b.c` for SDK constraints in version listing API

### ISSUE THREAD

**Issue by jonasfj** [dart-lang/pub-dev#4220]
We have [a check in `pub publish`][1] that ensures that the SDK constraint in `pubspec.yaml` does not use the caret-syntax (e.g. `^2.10.0`) for SDK constraints.

The comment there suggests that this is because some `pub` versions doesn't support caret-syntax. However, we could fix this on the server by rewriting `^a.b.c` to `>=a.b.c <a+1.b.c` (`Version.nextBreaking` from `pub_semver` helps doing this right; there is corner cases around `0.1.0` and possibly with prereleases).

And then update the [repository spec][2] to say that SDK constraints must be expressed in this format in the API. The constraint format inside the tarballs probably doesn't matter much.

**to do**:
 * [ ] Find latest versions of `pub` which doesn't support caret-syntax (probably Dart 1.x),
 * [ ] Investigate latest version of `pub` without caret-syntax support:
        * Can it handle current APIs, where `dependencies` may be using caret-syntax,
        * How does it handle SDK constraints with caret-syntax (what is the failure mode)
        * Is rewriting syntax in the API sufficient to make it work?
 * [ ] Rewrite `^a.b.c` to `>=a.b.c <a+1.b.c` for SDK constraints in the version listing API,
        * Handle unstable releases `0.a.b` correctly,
        * Handle pre-releases correctly (there may be corner cases wrt. `<a.b.c-dev` and `<a.b.c`)
 * [ ] Update `pub` such that usage of caret-syntax in `environment.sdk` is no-longer discouraged.

[1]: https://github.com/dart-lang/pub/blob/fb9ec4af9c2e64ebb30b8ddd985a1836a1f5c923/lib/src/validator/sdk_constraint.dart#L27-L40
[2]: https://github.com/dart-lang/pub/blob/master/doc/repository-spec-v2.md

**Comment by sigurdm** [comment]
I did a bisect of stable Dart SDK releases. First one with support for ^ syntax was 1.8.0.

**Comment by jonasfj** [comment]
Grepping through logs from past 60 days we find two unique IPs making requests against pub.dev with Dart versions less than 1.8.0. A total of 157 requests from two Dart versions less than 1.8.0.

Though that doesn't take into account Dart versions prior to introduction of the user-agent string in [71023003](https://codereview.chromium.org/71023003)

I guess we should look into what happens in Dart 1.7.0, if it find a caret constraint in the SDK constraints.
But it looks like it might be acceptable to break this few users.

**Comment by sigurdm** [comment]
I made a local mock-server that serves a single package with ^constraints.

pub 1.7.2:
```
> env PUB_HOSTED_URL=http://localhost:9999 pub get
Resolving dependencies... 
Error on line 1, column 1 of http://localhost:9999/api/packages/foo/versions/0.0.1: Invalid version constraint: Could not parse version "^2.1.0". Unknown text at "^2.1.0".
```
So an unpleasant crash (It doesn't just skip that version). If we were to keep support for old pub versions I think we need to do rewriting of the sdk constraints.

Interestingly, pub 1.7.2 seems to ignore the pubspec in the version listing `api/packages/foo`, but only parses the pubspec from `api/packages/foo/versions/0.0.1`.

**Comment by jonasfj** [comment]
> Given that pub would not download a package that doesn't resolve with the current sdk, I would think it would be enough to rewrite sdk constraints.

I agree, indeed it would probably be enough to only rewrite in the legacy end-point:
`api/packages/<package>/versions/<version>`
